### ISSUE TITLE
Create p5 instances from plain objects

### ISSUE THREAD

**Issue by Kinrany** [processing/p5.js#2978]
#### Nature of issue?

- [x] New feature request

#### Most appropriate sub-area of p5.js?

- [x] Core/Environment/Rendering

#### Details about the *missing feature*: 

- p5.js version: [0.6.1](https://github.com/processing/p5.js/tree/0.6.1)
- Steps to reproduce this: [codepen example](https://codepen.io/Kinrany/pen/dKMXeb)
   > TypeError: sketch is not a function

#### Feature details:

Use plain objects to create sketches in instance mode instead of a `(sketch) => {}` function that decorates it's argument with new methods.

Old way:

```javascript
new p5(sketch => {
  sketch.draw = () => {
    sketch.background(127);
  }
})
```

New way:

```javascript
new p5({
  draw(sketch) {
    sketch.background(127);
  }
});
```

---

For a more realistic example, suppose we have a sketch that
 - changes background color in a loop
 - draws a green circle wherever the mouse was clicked

```javascript
// https://codepen.io/Kinrany/pen/Yvwmqp

new p5(sketch => {
  
  // circles to be drawn
  let circles = [];
  
  sketch.setup = () => {
    sketch.createCanvas(200, 200);
  }
  
  sketch.mouseClicked = () => {
    // add one circle
    circles.push({
      x: sketch.mouseX,
      y: sketch.mouseY
    });
  }
  
  sketch.draw = () => {
    // change background color
    let time = sketch.millis() * 0.003;
    let red = 128 + Math.sin(time) * 128;
    let blue = 128 + Math.cos(time) * 128;
    sketch.background(red, 0, blue);
    
    // draw circles
    sketch.strokeWeight(0);
    sketch.fill(0, 255, 0);
    circles.forEach(c => sketch.ellipse(c.x, c.y, 10, 10));
  }
});
```

With this feature the sketch would look like this:

```javascript
// https://codepen.io/Kinrany/pen/qKZZrV

// circles to be drawn
let circles = [];

new_p5({
  setup(sketch) {
    sketch.createCanvas(200, 200);
  },

  mouseClicked(sketch) {
    // add one circle
    circles.push({
      x: sketch.mouseX,
      y: sketch.mouseY
    });
  },

  draw(sketch) {
    // change background color
    let time = sketch.millis() * 0.003;
    let red = 128 + Math.sin(time) * 128;
    let blue = 128 + Math.cos(time) * 128;
    sketch.background(red, 0, blue);

    // draw circles
    sketch.strokeWeight(0);
    sketch.fill(0, 255, 0);
    circles.forEach(c => sketch.ellipse(c.x, c.y, 10, 10));
  }
});
```

`new_p5` demo implementation:

```javascript
function new_p5(sketch_object) {
  return new p5(sketch => {
    // Convert each function on `sketch_object`
    // into a closure over `sketch`.
    // E.g. `sketch.draw = () => sketch_object.draw(sketch)`
    Object.entries(sketch_object).forEach(([name, method]) => {
      sketch[name] = () => method(sketch);
    });
  });
}
```

---

This way functions can conveniently use destructuring:

```javascript
new p5({
  draw({ ellipse }) {
    ellipse(30, 30, 10, 10);
    ellipse(30, 40, 10, 10);
    ellipse(40, 30, 10, 10);
    ellipse(40, 40, 10, 10);
  }
});
```

More importantly, sketches defined as objects are easier to consume or transform, and I struggle to find any downsides.

My use case is a reactive Vue component. It would be nice to be able to swap the sketch being displayed without restarting.

If there's interest in this, I'll try and submit a PR in a few days.

**Comment by meiamsome** [comment]
Destructuring like that in your example wouldn't work without the ellipse function being bound to the sketch object (I don't think it is in current p5 but I might be wrong).
Even if this wasn't the case I personally don't see any benefit over just doing the following:
```javascript
new p5(sketch => {
  const { ellipse } = sketch;
  sketch.draw = () => {
    ellipse(30, 30, 10, 10);
    ellipse(30, 40, 10, 10);
    ellipse(40, 30, 10, 10);
    ellipse(40, 40, 10, 10);
  };
});
```

Regarding your use case, can you elaborate further? What is preventing a usage of `remove` and then creating a new instance, or why would this instantiation be simpler?

Perhaps we should also consider permitting an object return from the function:
```javascript
new p5(sketch => {
  const { ellipse } = sketch;
  return {
    draw() {
      ellipse(30, 30, 10, 10);
      ellipse(30, 40, 10, 10);
      ellipse(40, 30, 10, 10);
      ellipse(40, 40, 10, 10);
    },
  };
});
```

I worry that passing an object literal would imply to users that they can use `this`, but that is not the case unless we bind functions onto the p5 instance (Which would not be terrible). To use your example:

```javascript
new_p5({
  circles: [],
  setup() {
    this.createCanvas(200, 200);
  },

  mouseClicked() {
    // add one circle
    this.circles.push({
      x: this.mouseX,
      y: this.mouseY
    });
  },

  draw() {
    // change background color
    let time = this.millis() * 0.003;
    let red = 128 + Math.sin(time) * 128;
    let blue = 128 + Math.cos(time) * 128;
    this.background(red, 0, blue);

    // draw circles
    this.strokeWeight(0);
    this.fill(0, 255, 0);
    this.circles.forEach(c => this.ellipse(c.x, c.y, 10, 10));
  },
});
```

In my opinion this is pretty readable. However, in doing so we would actually support the better looking (in my opinion, at least) class extension:

```javascript
class MySketch extends p5 {
  setup() {
    this.circles = [];
    this.createCanvas(200, 200);
  }

  mouseClicked() {
    // add one circle
    this.circles.push({
      x: this.mouseX,
      y: this.mouseY
    });
  }

  draw() {
    // change background color
    let time = this.millis() * 0.003;
    let red = 128 + Math.sin(time) * 128;
    let blue = 128 + Math.cos(time) * 128;
    this.background(red, 0, blue);

    // draw circles
    this.strokeWeight(0);
    this.fill(0, 255, 0);
    this.circles.forEach(c => this.ellipse(c.x, c.y, 10, 10));
  }
}
new MySketch();
```

**Comment by Kinrany** [comment]
I might have overestimated the convenience of destructiring:

```javascript
// default
sk.keyPressed = () => console.log(sk.keyCode)

// plain object, destructuring
keyPressed: ({ keyCode }) => console.log(keyCode)

// plain object, no destructuring
keyPressed: sk => console.log(sk.keyCode)
```

I guess it's only good for long functions, in which case one more line of code is less significant:

```javascript
// default
sk.draw = () => {
  let { line, mouseX, mouseY, pmouseX, pmouseY } = sk;
  // do something
}

// plain object
draw: ({ line, mouseX, mouseY, pmouseX, pmouseY }) => {
  // do something
}
```

And you're right that destructuring does not currently work with functions. Might be worth considering as a separate issue. It would allow writing sketches as if everything was global:

```javascript
new p5(sketch => {

  const { createCanvas, mouseX, mouseY, millis, background, strokeWeight, fill, ellipse } = sketch;
  
  // circles to be drawn
  let circles = [];
  
  function setup() {
    createCanvas(200, 200);
  }
  
  function mouseClicked() => {
    // add one circle
    circles.push({
      x: mouseX,
      y: mouseY
    });
  }
  
  function draw() {
    // change background color
    let time = millis() * 0.003;
    let red = 128 + Math.sin(time) * 128;
    let blue = 128 + Math.cos(time) * 128;
    background(red, 0, blue);
    
    // draw circles
    strokeWeight(0);
    fill(0, 255, 0);
    circles.forEach(c => ellipse(c.x, c.y, 10, 10));
  }
  
  Object.assign(sketch, { setup, mouseClicked, draw });
});
```

**Comment by Kinrany** [comment]
My use case: [vue-p5](https://github.com/Kinrany/vue-p5), an attempt to write a generic Vue component that contains a p5 instance.

Using the same example, this is how it works now:

```javascript
// https://codepen.io/Kinrany/pen/XYdEBx

const SketchContainer = {
  template: "<div/>",
  mounted() {
    new p5(sketch => {
      // simplified for demo purposes
      sketch.setup = () => this.$emit("setup", sketch);
      sketch.draw = () => this.$emit("draw", sketch);
      sketch.mouseClicked = () => this.$emit("mouse-clicked", sketch);
    }, this.$el);
  }
};

new Vue({
  el: "#app",
  template: `
    <vue-p5 
      @setup="setup" 
      @draw="draw"
      @mouse-clicked="mouseClicked">
    </vue-p5>`,
  components: {
    "vue-p5": SketchContainer
  },
  data: {
    circles: []
  },
  methods: {
    setup(sketch) {
      sketch.createCanvas(200, 200);
    },
    mouseClicked(sketch) {
      // add one circle
      this.circles.push({
        x: sketch.mouseX,
        y: sketch.mouseY
      });
    },
    draw(sketch) {
      // change background color
      let time = sketch.millis() * 0.003;
      let red = 128 + Math.sin(time) * 128;
      let blue = 128 + Math.cos(time) * 128;
      sketch.background(red, 0, blue);

      // draw circles
      sketch.strokeWeight(0);
      sketch.fill(0, 255, 0);
      this.circles.forEach(c => sketch.ellipse(c.x, c.y, 10, 10));
    }
  }
});
```

The component emits an event whenever the underlying p5 instance calls a sketch event.

This is not ideal: porting an existing sketch to Vue requires rewriting it. Ideally using the component should be exactly the same as calling `new p5` with one argument:

```html
<vue-p5 :sketch="sketch"></vue-p5>
```

I *could* just restart the sketch every time:

```javascript
// https://codepen.io/Kinrany/pen/XYdEBx

const SketchContainer = {
  props: ["sketch"],
  template: "<div/>",
  data: () => ({
    instance: null
  }),
  mounted() {
    this.instance = new p5(this.sketch, this.$el);
  },
  watch: {
    sketch(newSketch) {
      this.instance.remove();
      this.instance = new p5(newSketch, this.$el);
    }
  }
};

new Vue({
  el: "#app",
  template: `<vue-p5 :sketch="sketch"></vue-p5>`,
  components: {
    "vue-p5": SketchContainer
  },
  data: {
    sketch: sketch => {
      // no changes here
    }
  }
});
```

But that would make writing idiomatic Vue code with it somewhat awkward:

```javascript
// https://codepen.io/Kinrany/pen/jKqpzL

const SketchContainer = {
  // no changes here
};

new Vue({
  // no changes here

  data: {
    // circles to be drawn
    circles: []
  },
  methods: {
    sketch(sketch) {
      sketch.setup = () => this.setup(sketch);
      sketch.mouseClicked = () => this.mouseClicked(sketch);
      sketch.draw = () => this.draw(sketch);
    },
    setup(sketch) {
      sketch.createCanvas(200, 200);
    },
    mouseClicked(sketch) {
      // add one circle
      this.circles.push({
        x: sketch.mouseX,
        y: sketch.mouseY
      });
    },
    draw(sketch) {
      // change background color
      let time = sketch.millis() * 0.003;
      let red = 128 + Math.sin(time) * 128;
      let blue = 128 + Math.cos(time) * 128;
      sketch.background(red, 0, blue);

      // draw circles
      sketch.strokeWeight(0);
      sketch.fill(0, 255, 0);
      this.circles.forEach(c => sketch.ellipse(c.x, c.y, 10, 10));
    }
  }
});
```

Basically the same as the first example, just worse. 

So currently I see three options, and each has a downside:
 - existing sketches cannot be used
 - idiomatic Vue sketches have extra boilerplate
 - container component has two implementations in parallel

With plain objects it would be just this:

```javascript
const SketchContainer = {
  props: ["sketch"],
  template: "<div/>",
  mounted() {
    new p5({
      // simplified for demo purposes
      setup: sk => this.sketch.setup(sk),
      draw: sk => this.sketch.draw(sk),
      mouseClicked: sk => this.sketch.mouseClicked(sk)
    });
  }
};

new Vue({
  el: "#app",
  template: `<vue-p5 :sketch="this"></vue-p5>`,
  
  // `sketch` method removed, no other changes
});
```

So the component can be used with both plain objects and Vue component sketches.

There is no reason to support Vue specifically, of course, but I think this demonstrates well that with objects it's easier to create new p5-compatible implementations.

**Comment by Kinrany** [comment]
Does p5 use `this`, inheritance and classes anywhere in it's API?

Personally I'd rather use a parameter. It's explicit, allows arrow functions and doesn't need `var that = this`.

Permitting object return is an interesting idea. With it one could use instance mode as if it was global mode:

```javascript
new p5(sketch => {
  Object.assign(window, sketch);
  function draw() { 
    background(200);
  }
  return { draw };
}, some_html_element);
```

It doesn't help with transforming the sketch though. The argument is still a closure over `sketch`, so one would have to use Proxy hacks to use it in a different context.

**Comment by Kinrany** [comment]
On second thought posting a wall of ~~text~~ code that uses a not very popular framework is not a good idea D:

Is there anything I can do to make all that easier to read?

**Comment by Kinrany** [comment]
I'm trying to come up with a case in plain JS, but using Proxy is always an option, there's nothing I can't do with sketches the way they are now, so changing p5 is not necessary. 

I guess I just don't like sketches-as-decorators, so I'll close this issue for now. But please feel free to @ me if you have any thoughts on this topic.

**Comment by kenny1983** [comment]
> However, in doing so we would actually support the better looking (in my opinion, at least) class extension

That's my opinion too @meiamsome, and I've been looking for a way to implement such beautiful (and modern) code! Is this something that's actually been suggested before, or should I do the honours?
